ucfg_test.go

```diff
@@ -61,3 +68,284 @@ func TestFlattenKeys(t *testing.T) {
 		})
 	}
 }
+
+func TestDetectDirectCyclicReference(t *testing.T) {
+	tests := []struct {
+		title    string
+		cfg      map[string]interface{}
+		expected string
+	}{
+		{
+			title: "direct cyclic reference",
+			cfg: map[string]interface{}{
+				"top.reference": "${top.reference}",
+			},
+		},
+		{
+			title: "compound reference",
+			cfg: map[string]interface{}{
+				"top.reference": "hello ${top.reference}",
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.title, func(t *testing.T) {
+			c, err := NewFrom(test.cfg, opts...)
+
+			config := &struct {
+				TopReference string `config:"top.reference"`
+			}{}
+
+			err = c.Unpack(&config, opts...)
+			assert.Error(t, err)
+		})
+	}
+}
+
+func TestDetectIndirectCyclicReference(t *testing.T) {
+	cfg := map[string]interface{}{
+		"a": map[string]interface{}{
+			"b": map[string]interface{}{
+				"c": "${a}",
+			},
+		},
+	}
+
+	config := struct {
+		A struct {
+			B struct {
+				C map[string]interface{}
+			}
+		}
+	}{}
+
+	c, err := NewFrom(cfg, opts...)
+	err = c.Unpack(&config, opts...)
+	if assert.Error(t, err) {
+		assert.Equal(t, "cyclic reference detected for key: 'a' accessing 'a.b'", err.Error())
+	}
+}
+
+func TestConfigCyclicReference(t *testing.T) {
+	cfg := map[string]interface{}{
+		"a": map[string]interface{}{
+			"b": map[string]interface{}{
+				"c": "${a}",
+			},
+		},
+	}
+
+	config := struct {
+		A struct {
+			B struct {
+				C *Config
+			}
+		}
+	}{}
+
+	c, err := NewFrom(cfg, opts...)
+	err = c.Unpack(&config, opts...)
+	assert.NoError(t, err)
+}
+
+func TestDetectArrayCyclicReference(t *testing.T) {
+	cfg := map[string]interface{}{
+		"c": []interface{}{
+			"${c.0}",
+			"value 2",
+		},
+	}
+
+	config := &struct {
+		C []string `config:"c"`
+	}{}
+
+	c, err := NewFrom(cfg, opts...)
+	assert.NoError(t, err)
+
+	err = c.Unpack(&config, opts...)
+	if assert.Error(t, err) {
+		assert.Equal(t, "cyclic reference detected for key: 'c.0' accessing 'c.0'", err.Error())
+	}
+}
+
+func TestCyclicReferenceShouldFallbackToOtherResolvers(t *testing.T) {
+	cfg := map[string]interface{}{
+		"top.reference": "${top.reference}",
+	}
+
+	resolveFn := func(key string) (string, error) {
+		if key == "top.reference" {
+			return "reference-found", nil
+		}
+		return "", ErrMissing
+	}
+
+	opts := []Option{
+		PathSep("."),
+		Resolve(resolveFn),
+		ResolveEnv,
+		VarExp,
+	}
+
+	c, err := NewFrom(cfg, opts...)
+	v, err := c.String("top.reference", -1, opts...)
+	if assert.NoError(t, err) {
+		assert.Equal(t, "reference-found", v)
+	}
+}
+
+func TestTopYamlKeyInEnvResolvers(t *testing.T) {
+	resolveFn := func(key string) (string, error) {
+		if key == "a.key" {
+			return "key-found", nil
+		}
+		return "", fmt.Errorf("could not find the key: %s", key)
+	}
+
+	opts := []Option{
+		PathSep("."),
+		Resolve(resolveFn),
+		ResolveEnv,
+		VarExp,
+	}
+
+	tests := []struct {
+		name     string
+		cfg      map[string]interface{}
```

Comment:
tip: `cfg interface{}`. Then you can also use `&struct{ ...} `, to pass anonymous structs. Checkout the `TestUnpackX` tests.